### PR TITLE
perf: Optimize `is_sorted` for numeric data

### PR DESCRIPTION
**Pull request by ritchie46**
Should be much faster and failing fast in case of non-sorted data.

### CODE REVIEW

File: crates/polars-time/src/chunkedarray/rolling_window/dispatch.rs [hunk 1]
```diff
@@ -82,8 +82,9 @@ where
     if ca.is_empty() {
         return Ok(Series::new_empty(ca.name(), ca.dtype()));
     }
-    let ca = ca.rechunk();
-    let by = by.rechunk();
+    if by.null_count() > 0 {
```
Comment (ritchie46) [inline crates/polars-time/src/chunkedarray/rolling_window/dispatch.rs:85]: This could be factored out of branch and be checked before any compute (rechunking).

File: crates/polars-time/src/chunkedarray/rolling_window/dispatch.rs [hunk 2]
```diff
@@ -98,15 +99,18 @@ where
...
+    let by_is_sorted = by.is_sorted(SortOptions {
+        descending: false,
+        ..Default::default()
+    })?;
     let by = by.datetime().unwrap();
     let tu = by.time_unit();
 
     let func = rolling_agg_fn_dynamic;
-    let out: ArrayRef = if matches!(by.is_sorted_flag(), IsSorted::Ascending) {
+    let out: ArrayRef = if by_is_sorted {
```
Comment (ritchie46) [inline crates/polars-time/src/chunkedarray/rolling_window/dispatch.rs:112]: This checks the sortedness flags and otherwise does a fast sortedness check that fails fast on non-sorted arrays.

**Review by ritchie46 — COMMENTED** [review]
@MarcoGorelli check comments FYI.

File: crates/polars-time/src/chunkedarray/rolling_window/dispatch.rs [hunk 3]
```diff
@@ -82,8 +82,9 @@ where
     if ca.is_empty() {
         return Ok(Series::new_empty(ca.name(), ca.dtype()));
     }
-    let ca = ca.rechunk();
-    let by = by.rechunk();
+    if by.null_count() > 0 {
+        polars_bail!(InvalidOperation: "'Expr.rolling_*_by(...)' not yet supported for series with null values, consider using 'DataFrame.rolling' or 'Expr.rolling'")
+    }
```
Comment (MarcoGorelli) [inline crates/polars-time/src/chunkedarray/rolling_window/dispatch.rs:87]: this needs checking for `ca` too? looks like it's being removed from L107 but not replaced

Comment (MarcoGorelli) [reply crates/polars-time/src/chunkedarray/rolling_window/dispatch.rs:87]: also, L114 can probably now just become
```rust
        let by_values = by.cont_slice().expect("`by` has already been rechunked and checked for null values");
```
?

likewise L137

Comment (ritchie46) [reply crates/polars-time/src/chunkedarray/rolling_window/dispatch.rs]: It isn't I believe, but we should check that higher up indeed. :+1:

Comment (ritchie46) [reply crates/polars-time/src/chunkedarray/rolling_window/dispatch.rs]: Ah, it is. NVM.

File: crates/polars-time/src/chunkedarray/rolling_window/dispatch.rs [hunk 4]
```diff
@@ -130,15 +132,7 @@ where
...
-        if arr.null_count() > 0 {
-            polars_bail!(InvalidOperation: "'Expr.rolling_*_by(...)' not yet supported for series with null values, consider using 'DataFrame.rolling' or 'Expr.rolling'")
-        }
-        let by_values = by.cont_slice().map_err(|_| {
-            polars_err!(
-                ComputeError:
-                "`by` column should not have null values in 'rolling by' expression"
-            )
-        })?;
+        let by_values = by.cont_slice().expect("checked already");
```
Comment (MarcoGorelli) [inline crates/polars-time/src/chunkedarray/rolling_window/dispatch.rs:135]: same in L112 - I can do this in a separate PR if you like though, where I'll add tests to check that this raises on null values - it's a bit concerning that in the previous commit the tests passed
